_posts/2023-11-29-figgy-migration-details.md

```diff
@@ -0,0 +1,153 @@
+- Check row counts for all tables between both servers:
+```sql
+WITH tbl AS
+  (SELECT table_schema,
+          TABLE_NAME
+   FROM information_schema.tables
+   WHERE TABLE_NAME not like 'pg_%'
+     AND table_schema in ('public'))
+SELECT table_schema,
+       TABLE_NAME,
+       (xpath('/row/c/text()', query_to_xml(format('select count(*) as c from %I.%I', table_schema, TABLE_NAME), FALSE, TRUE, '')))[1]::text::int AS rows_n
+FROM tbl
+ORDER BY rows_n DESC;
+```
+- Assuming they're close, move on. If they're far off, stop here. Something's broken.
+- Notify your users. We did this using Slack: `@here We're now starting scheduled maintenance of Figgy, it will be going into read only mode for up to 3 hours. We'll send another message here when it's ready for regular use again.`
+- Set your application to read-only mode (or shut it down), so nobody can add new records while you migrate.
+- Check row counts for all tables between both servers again (belt and suspenders, folks).
+- If they're not exact, wait a minute, try again. If they never become exact, then stop. Otherwise, continue.
+- [Update the sequences](https://wiki.postgresql.org/wiki/Fixing_Sequences) in the new database.
+- Run the following query in the new database's psql terminal:
+
+```sql
+SELECT 
+    'SELECT SETVAL(' ||
+       quote_literal(quote_ident(sequence_namespace.nspname) || '.' || quote_ident(class_sequence.relname)) ||
+       ', COALESCE(MAX(' ||quote_ident(pg_attribute.attname)|| '), 1) ) FROM ' ||
+       quote_ident(table_namespace.nspname)|| '.'||quote_ident(class_table.relname)|| ';'
+FROM pg_depend 
+    INNER JOIN pg_class AS class_sequence
+        ON class_sequence.oid = pg_depend.objid 
+            AND class_sequence.relkind = 'S'
+    INNER JOIN pg_class AS class_table
+        ON class_table.oid = pg_depend.refobjid
+    INNER JOIN pg_attribute 
+        ON pg_attribute.attrelid = class_table.oid
+            AND pg_depend.refobjsubid = pg_attribute.attnum
+    INNER JOIN pg_namespace as table_namespace
+        ON table_namespace.oid = class_table.relnamespace
+    INNER JOIN pg_namespace AS sequence_namespace
+        ON sequence_namespace.oid = class_sequence.relnamespace
+ORDER BY sequence_namespace.nspname, class_sequence.relname;
+```
+- This will output a new query which you should copy and paste and run.
+
+- Update the application configuration to point to the new database.
+- Restart your web server and any workers.
+- Make sure your test page loads.
+- We have multiple servers in a load-balanced setup, so we did those three steps multiple times. We removed half of our servers from the load balancer, updated them, reinstated them, removed the other half from the load balancer, restarted the web servers and workers, then checked that our test object still loaded before repeating those steps on our other servers. This gave us a way to retreat if we needed to. It also kept the system available for read operations during the upgrade.
```

Comment:
"those three steps" is sort of confusing since the steps aren't numbered. What if we say "we updated the application server configuration in two batches"
```suggestion
- We have multiple servers in a load-balanced setup, so we updated the application server configuration in two batches. We removed half of our servers from the load balancer, updated them, reinstated them, removed the other half from the load balancer, restarted the web servers and workers, then checked that our test object still loaded before repeating those steps on our other servers. This gave us a way to retreat if we needed to. It also kept the system available for read operations during the upgrade.
```